### PR TITLE
Use Dart 3 patterns and records in `adaptive_app` codelab

### PR DESCRIPTION
**Pull request by miquelbeltran**
Using Patterns and Records from Dart 3 in the adaptive_app codelab.

Googlers: cl/533649426

## Pre-launch Checklist

- [x] I read the [Effective Dart: Style] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-devrel channel on [Discord].

[Effective Dart: Style]: https://dart.dev/guides/language/effective-dart/style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: adaptive_app/step_07/lib/src/adaptive_playlists.dart [hunk 1]
```diff
@@ -76,14 +77,11 @@ class _WideDisplayPlaylistsState extends State<WideDisplayPlaylists> {
...
-                playlistName: selectedPlaylist!.snippet!.title!)
-          else
-            const Center(
-              child: Text('Select a playlist'),
-            ),
+          switch ((selectedPlaylist?.id, selectedPlaylist?.snippet?.title)) {
+            (String id, String title) =>
+              PlaylistDetails(playlistId: id, playlistName: title),
+            _ => const Center(child: Text('Select a playlist')),
+          },
```
Comment (miquelbeltran) [inline adaptive_app/step_07/lib/src/adaptive_playlists.dart:84]: Similar to the previous example, in this case it uses a record (id, title) to match against non-null values. Maybe I am trying too hard, but the advantage vs. the old one is that there's no forced cast to non-null.

**Comment by miquelbeltran** [comment]
Before applying all the changes, I'd like to know what you think @domesticmouse @johnpryan 

I'm a bit afraid that using patterns in the two cases I highlight may make things too complex

File: adaptive_app/step_07/lib/src/adaptive_playlists.dart (hunk 1 above)
Comment (johnpryan) [reply adaptive_app/step_07/lib/src/adaptive_playlists.dart:84]: This is a great example of a record pattern, records are handy as short-lived objects like this, and we don't have many examples that show how record patterns can be used.

File: adaptive_app/step_07/lib/src/adaptive_playlists.dart [hunk 2]
```diff
@@ -64,9 +64,10 @@ class _WideDisplayPlaylistsState extends State<WideDisplayPlaylists> {
   Widget build(BuildContext context) {
     return Scaffold(
       appBar: AppBar(
-        title: selectedPlaylist == null
-            ? const Text('Your Playlists')
-            : Text('Your Playlist: ${selectedPlaylist!.snippet!.title!}'),
+        title: switch (selectedPlaylist?.snippet?.title) {
+          String title => Text('Your Playlist: $title'),
+          _ => const Text('Your Playlists'),
+        },
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply adaptive_app/step_07/lib/src/adaptive_playlists.dart:70]: I like it, a switch expression is more readable than the ternary operator for me, but I've been seeing a lot of switch expressions lately.

**Review by johnpryan — APPROVED** [review]

**Comment by miquelbeltran** [comment]
thanks for the feedback! I will apply the changes retroactively to the codelab and update the document

File: adaptive_app/step_07/lib/src/adaptive_playlists.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (domesticmouse) [reply adaptive_app/step_07/lib/src/adaptive_playlists.dart:70]: I concur, `switch` feels more readable to me here as well. It'll take a while for people to adapt to the syntax, but I really like the declarative nature of the pattern matching approach.

File: adaptive_app/step_07/lib/src/adaptive_playlists.dart (hunk 1 above)
Comment (domesticmouse) [reply adaptive_app/step_07/lib/src/adaptive_playlists.dart:84]: Yup, concur.

**Review by domesticmouse — APPROVED** [review]
LGTM

**Comment by miquelbeltran** [comment]
@domesticmouse the codelab document "Adaptive Apps in Flutter" has been updated as well

**Comment by domesticmouse** [comment]
> @domesticmouse the codelab document "Adaptive Apps in Flutter" has been updated as well

I'll regenerate the codelab and land this PR when that CL is published.

**Comment by domesticmouse** [comment]
Currently I can't regenerate this codelab, apparently it has too many images to export it from docs. 

Le sigh.

**Pull request merged**